service/worker/deletenamespace/deleteexecutions/workflow.go

```diff
@@ -0,0 +1,194 @@
+	"go.temporal.io/server/service/worker/deletenamespace/defaults"
+	"go.temporal.io/server/service/worker/deletenamespace/errors"
+)
+
+const (
+	WorkflowName = "temporal-sys-delete-executions-workflow"
+
+	heartbeatEveryExecutions = 1000
+)
+
+type (
+	DeleteExecutionsParams struct {
+		Namespace   namespace.Name
+		NamespaceID namespace.ID
+
+		DeleteRPS                                 int
+		ListPageSize                              int
+		ConcurrentDeleteExecutionsActivitiesCount int
+
+		PreviousSuccessCount int
+		PreviousErrorCount   int
+	}
+
+	DeleteExecutionsResult struct {
+		SuccessCount int
+		ErrorCount   int
+	}
+)
+
+var (
+	retryPolicy = &temporal.RetryPolicy{
+		InitialInterval: 1 * time.Second,
+		MaximumInterval: 10 * time.Second,
+	}
+
+	localActivityOptions = workflow.LocalActivityOptions{
+		RetryPolicy:            retryPolicy,
+		StartToCloseTimeout:    10 * time.Second,
+		ScheduleToCloseTimeout: 5 * time.Minute,
+	}
+
+	deleteWorkflowExecutionsActivityOptions = workflow.ActivityOptions{
+		RetryPolicy:            retryPolicy,
+		StartToCloseTimeout:    60 * time.Minute,
+		ScheduleToCloseTimeout: 6 * time.Hour,
+	}
+)
+
+func validateParams(params *DeleteExecutionsParams) error {
+	if params.NamespaceID.IsEmpty() {
+		return temporal.NewNonRetryableApplicationError("namespace ID is empty", "", nil)
+	}
+
+	if params.Namespace.IsEmpty() {
+		return temporal.NewNonRetryableApplicationError("namespace is empty", "", nil)
+	}
+
+	if params.DeleteRPS <= 0 {
+		params.DeleteRPS = defaults.DeleteRPS
+	}
+	if params.ListPageSize <= 0 {
+		params.ListPageSize = defaults.ListPageSize
+	}
+	if params.ConcurrentDeleteExecutionsActivitiesCount <= 0 {
+		params.ConcurrentDeleteExecutionsActivitiesCount = defaults.ConcurrentDeleteExecutionsActivitiesCount
+	}
+	if params.ConcurrentDeleteExecutionsActivitiesCount > defaults.MaxConcurrentDeleteExecutionsActivitiesCount {
+		params.ConcurrentDeleteExecutionsActivitiesCount = defaults.MaxConcurrentDeleteExecutionsActivitiesCount
+	}
+	return nil
+}
+
+func DeleteExecutionsWorkflow(ctx workflow.Context, params DeleteExecutionsParams) (DeleteExecutionsResult, error) {
+	logger := workflow.GetLogger(ctx)
+	logger.Info("Child workflow started.", tag.WorkflowType(WorkflowName))
+	result := DeleteExecutionsResult{
+		SuccessCount: params.PreviousSuccessCount,
+		ErrorCount:   params.PreviousErrorCount,
+	}
+
+	if err := validateParams(&params); err != nil {
+		return result, err
+	}
+
+	// Minimum value is heartbeatEveryExecutions/params.DeleteRPS. 2 is "make sure" multiplicator.
+	deleteWorkflowExecutionsActivityOptions.HeartbeatTimeout = time.Duration(heartbeatEveryExecutions/params.DeleteRPS*2) * time.Second
+
+	var a *Activities
+	var nextPageToken []byte
+	runningDeleteExecutionsActivityCount := 0
+	runningDeleteExecutionsSelector := workflow.NewSelector(ctx)
+	var lastDeleteExecutionsActivityErr error
+
+	for i := 0; i < params.ConcurrentDeleteExecutionsActivitiesCount; i++ {
```

Comment:
So this would start concurrent number of activities, and wait all of them to complete then continue as new. If that is what you want, you don't need the selector and all those controls. You just need a for loop that simply start all activities, put future in a slice, and then another for loop to wait on each of them.
But I think that is not ideal, you want another control of PageCountPerExecution which controls how many pages of WorkflowExecutions you want to handle before continue as new.

Reply:
This is very good point. I intended to do the later but did it wrong.